Simplicity descriptors

uncomputable opened the pull request:
Takes ideas from https://github.com/BlockstreamResearch/rust-simplicity/pull/154 and implements Simplicity descriptors as a special kind of script-spend inside a Taproot descriptor. Simplicity and Miniscript can be mixed in the same tree. `TapLeafScript` is an interface that unifies both worlds. All methods use this interface to work as normal.

apoelstra commented:
Some high-level comments:

* Why do the iterators panic? `TapTreeIter` should just be able to yield both kinds of leaves.
* Why are we restricted to only one Simplicity leaf exactly? This isn't really explained in the PR.

uncomputable commented:
I generalized Simplicity descriptors to multiple leaves which might be mixed with Miniscript.

uncomputable commented:
Rebased, bumped the MSRV to fix CI, and fixed the sub crates that silently broke.

https://github.com/BlockstreamResearch/rust-simplicity/pull/170 adds the required changes on the Simplicity side. We need to merge that first, but the Miniscript side dictates what we need on the Simplicity side. So we need to look at both PRs.

Right now the control block is broken somehow. I cannot spend UTXOs with my wallet. I will investigate.

apoelstra commented:
I don't think we need to merge rust-simplicity 170 first -- we should merge that, update the vendored copy of simplicity, and cut a new release, but for now here we can just keep pointing to your branch in our Cargo.toml. We can fix that in a later PR.

apoelstra commented on src/descriptor/tr.rs:
```diff
@@ -136,12 +138,30 @@ impl<Pk: MiniscriptKey, Ext: Extension> TapTree<Pk, Ext> {
     {
+        struct SimTranslator<'a, T>(&'a mut T);
+
+        impl<'a, Pk, T, Q, Error> simplicity::Translator<Pk, Q, Error> for SimTranslator<'a, T>
```
In 899269acefeb534f140cfd4d7f86b66ad567b55f:

I wonder if, in rust-simplicity, we could just implement `SimTranslator` for all `Translator`s. Or if that'd break stuff.

uncomputable replied:
`elements_miniscript::Translator` and `miniscript::Translator` are distinct for some reason. If this doesn't clash with future plans, we could use `miniscript::Translator` across `elements_miniscript`, and implement `simplicity::Translator` for every `miniscript::Translator`.

I tried it out and the code compiled on my end.

apoelstra commented on src/descriptor/tr.rs:
```diff
@@ -263,7 +286,7 @@ impl<Pk: MiniscriptKey, Ext: Extension> Tr<Pk, Ext> {
     pub fn sanity_check(&self) -> Result<(), Error> {
         for (_depth, ms) in self.iter_scripts() {
-            ms.sanity_check()?;
+            ms.as_miniscript().unwrap().sanity_check()?;
```
In 899269acefeb534f140cfd4d7f86b66ad567b55f:

I'd be tempted to just return `false` here rather than panicking. Though I'd leave a FIXME comment.

apoelstra replied:
Ah, never mind, this is replaced in a later commit (and has a TODO comment on it there to implement it more fully).

apoelstra commented on src/descriptor/tr.rs:
```diff
@@ -244,7 +267,7 @@ impl<Pk: MiniscriptKey, Ext: Extension> Tr<Pk, Ext> {
             let mut builder = TaprootBuilder::new();
             for (depth, ms) in self.iter_scripts() {
-                let script = ms.encode();
+                let script = ms.as_miniscript().unwrap().encode();
```
In 899269acefeb534f140cfd4d7f86b66ad567b55f:

Can you add a FIXME comment here so we remember to fix it?

apoelstra replied:
Ah, nope, fixed in a later commit.

apoelstra commented on src/descriptor/tr.rs:
```diff
@@ -338,6 +362,7 @@ impl<Pk: MiniscriptKey, Ext: Extension> Tr<Pk, Ext> {
 
         tree.iter()
             .filter_map(|(depth, ms)| {
+                let ms = ms.as_miniscript().unwrap();
```
In 899269acefeb534f140cfd4d7f86b66ad567b55f: add a FIXME

apoelstra replied:
Never mind, fixed later (with an error return and a TODO).

apoelstra commented on src/descriptor/tr.rs:
```diff
@@ -293,6 +316,7 @@ impl<Pk: MiniscriptKey, Ext: Extension> Tr<Pk, Ext> {
 
         tree.iter()
             .filter_map(|(depth, ms)| {
+                let ms = ms.as_miniscript().unwrap();
```
In 899269acefeb534f140cfd4d7f86b66ad567b55f: add a FIXME

apoelstra replied:
Never mind, fixed later.

apoelstra commented on src/descriptor/tr.rs:
```diff
@@ -650,7 +711,7 @@ impl<Pk: MiniscriptKey, Ext: Extension> ForEachKey<Pk> for Tr<Pk, Ext> {
         let script_keys_res = self
             .iter_scripts()
-            .all(|(_d, ms)| ms.for_each_key(&mut pred));
+            .all(|(_d, ms)| ms.as_miniscript().unwrap().for_each_key(&mut pred));
```
In 899269acefeb534f140cfd4d7f86b66ad567b55f: add a FIXME

apoelstra replied:
Ah, nope, this is replaced in a later commit.

apoelstra commented on src/descriptor/tr.rs:
```diff
@@ -729,6 +790,7 @@ where
         // map (lookup_control_block) from the satisfier here.
         let (mut min_wit, mut min_wit_len) = (None, None);
         for (depth, ms) in desc.iter_scripts() {
+            let ms = ms.as_miniscript().unwrap();
```
In 899269acefeb534f140cfd4d7f86b66ad567b55f: add a FIXME

apoelstra commented on src/psbt/mod.rs:
```diff
@@ -1290,6 +1290,9 @@ where
         debug_assert_eq!(_depth_der, depth);
+        let ms_derived = ms_derived.as_miniscript().unwrap();
+        let ms = ms.as_miniscript().unwrap();
+
```
In 899269acefeb534f140cfd4d7f86b66ad567b55f:

This one we should return an error on. I think we're a long way away from supporting PSBT/PSET with Simplicity policies. Though we should also add a FIXME.

apoelstra replied:
Nope, fixed later :)

apoelstra commented on src/descriptor/tr.rs:
```diff
@@ -623,6 +683,7 @@ impl<Pk: MiniscriptKey, Ext: Extension> Liftable<Pk> for TapTree<Pk, Ext> {
                     Ok(Policy::Threshold(1, vec![lift_helper(l)?, lift_helper(r)?]))
                 }
                 TapTree::Leaf(ref leaf) => leaf.lift(),
+                TapTree::SimplicityLeaf(..) => panic!("Cannot lift Simplicity policy to Miniscript semantic policy"),
```
In 899269acefeb534f140cfd4d7f86b66ad567b55f:

Can you stick the word FIXME into this error message?

apoelstra replied:
This is the only panic added in the first commit which isn't removed by the end of the PR.

apoelstra reviewed (approved):
ACK 194edcfc81a0780a782dafc3092567fc4c6c823e

uncomputable commented:
Added the FIXME and squashed. I can also confirm that the descriptors work inside the Simplicity wallet if one uses the Simplicity version that is compatible with Elements.

apoelstra commented:
72fde205f45b2ee050e7309b893f92c749aa7173 does not compile.

00862f06547dfeeb915ac96d700fd5853d0e832f is good otherwise.

uncomputable commented:
Now everything should compile. Sorry for that. I should have checked.

apoelstra reviewed (approved):
ACK 2ff1a175311fa822c5ec9b323efcc850a34086e3
